advanced/sr_gui_fingertip_visualization/src/sr_gui_fingertip_visualization/visual_tab_layouts.py

```diff
@@ -0,0 +1,439 @@
+                self._data_labels[finger]['pac'][i] = QLabel("P{}:-".format(i))
+                remaining_data_layout_pac.addWidget(self._data_labels[finger]['pac'][i], current_row, current_column,
+                                                    alignment=Qt.AlignCenter)
+                current_column +=1
+                if current_column == columns:
+                    current_column = 0
+                    current_row += 1
+            '''
+            text_data_layout = QGridLayout()
+            self._data_labels[finger]['pac0'] = QLabel("pac0:-")
+            self._data_labels[finger]['pac1'] = QLabel("pac1:-")
+            self._data_labels[finger]['pdc'] = QLabel("pdc:-")
+            self._data_labels[finger]['tac'] = QLabel("tac:-")
+            self._data_labels[finger]['tdc'] = QLabel("tdc:-")
+
+            text_data_layout.addWidget(self._data_labels[finger]['pac0'], 0, 0, alignment=Qt.AlignCenter)
+            text_data_layout.addWidget(self._data_labels[finger]['pac1'], 0, 1, alignment=Qt.AlignCenter)
+            text_data_layout.addWidget(self._data_labels[finger]['pdc'], 0, 2, alignment=Qt.AlignCenter)
+            text_data_layout.addWidget(self._data_labels[finger]['tac'], 0, 3, alignment=Qt.AlignCenter)
+            text_data_layout.addWidget(self._data_labels[finger]['tdc'], 0, 4, alignment=Qt.AlignCenter)
+
+            layout.addLayout(text_data_layout)
+
+        elif detected_fingertip_type == BiotacType.SP_MINUS:
+            self._finger_widgets[finger] = DotUnitBiotacSPMinus(self._finger_frame[finger])
+            layout.addWidget(self._finger_widgets[finger], alignment=Qt.AlignCenter)
+
+        elif detected_fingertip_type == BiotacType.BLANK:
+            self._finger_widgets[finger] = QLabel("No tactile sensor")
+            layout.addWidget(self._finger_widgets[finger], alignment=Qt.AlignCenter)
+
+        self._finger_frame[finger].setLayout(layout)
+        return self._finger_frame[finger]
+
+    def _detect_biotac_type(self, finger):
+        sum_of_electrode_values = sum(self._data[finger]['electrodes'])
+        biotac_type = None
+        if sum_of_electrode_values == 0:
+            biotac_type = BiotacType.BLANK
+        elif sum_of_electrode_values == self._electrode_count:
+            biotac_type = BiotacType.SP_MINUS
+        else:
+            biotac_type = BiotacType.SP_PLUS
+        return biotac_type
+
+    def start_timer_and_subscriber(self):
+        self._subscriber = rospy.Subscriber('/{}/tactile'.format(self._side), BiotacAll, self._tactile_data_callback)
+        self._timer.timeout.connect(self.timerEvent)
+        self._timer.start(10)
+
+    def stop_timer_and_subscriber(self):
+        self._timer.stop()
+        if self._subscriber:
+            self._subscriber.unregister()
+
+    def timerEvent(self):
+        for finger in self._fingers:
+            if self._finger_frame[finger].isChecked():
+                if type(self._finger_widgets[finger]) == list:
```

Comment:
It is strongly not recommended to have == to test a type. A slightly better way is type(x) is list, but an even better is to use isinstance(x, list)